### DISCUSSION TITLE
New Keywords for Modal Reactors

### DISCUSSION THREAD

**Discussion by a-sr** [lf-lang/lingua-franca#947] [General]
With [modal reactors](#501) four new keywords were introduced: *mode*, *initial*, *reset*, *continue*. Since these then could no longer be used for other ids, such as variable names, it caused a few problems with existing models. This raised concerns for introducing new keywords, especially common ones, such as *reset*.

## Current Implementation in Modal Reactors

The *mode* keyword is essential for the parsing process, hence, there is no way around it but there were no models using it anyway.
*reset* and *continue* were initially an enum rule in the grammar. However, *reset* is used rather often as an input port in Savia benchmark models. Hence, I implemented @petervdonovan's idea and replaced the predefined enum filed by a string. As a consequence the grammar accepts any input at this location but with additional validator rules, code completion proposals, and semantics syntax highlighting, the old keyword behavior could be simulated. In contrast to the initially proposed syntax for mode transition effects on reactions, the transition modifier is now in parentheses to ease parsing:
```
reaction(t) -> ModeA, ModeB(continue), ModeB(reset) {=
```
The *initial* keyword could be implemented as a string as well, however, since it represents a boolean value and is currently located before the *mode* keyword, which is relevant for the parser look ahead (``(initial?='initial')? 'mode'``), I am not sure if it is beneficial to do this. Hence, *initial* remains a keyword for now. As a consequence, one [existing model](https://github.com/lf-lang/lingua-franca/commit/c9152ed6a55a0ba104942776802cddda67a89d9a) had to be adjusted, as well as a [target property](https://github.com/lf-lang/lingua-franca/commit/dfc05478bbf75bf7673baca933391f8c8574332e).

## Discussion Topics

* Is the syntax for transitions modifier (*reset*, *continue*) in accordance with other syntax constructs?
* Is necessary to also unblock the term *initial* by using the the string workaround?
* Do we want to change the target property syntax in order to prevent collisions with keywords, e.g. by introducing quotations marks?
* Do we want to establish a strategy for handling future keywords, i.e. when to use the string workaround?

**Comment by oowekyala** [comment]
Maybe an attribute (#756) would be appropriate for those? `@initial mode`, `@continue ModeB`. The advantage is they don't reserve keywords, but you'll need to validate them similar to strings.

**Comment by cmnrd** [comment]
> Is the syntax for transitions modifier (reset, continue) in accordance with other syntax constructs?

I don't think we currently have a syntax similar to your proposal. What immediately comes to my mind is the `interleaved` modifier on ports, but here the syntax works the other way around: `interleaved(bank.port)`. In your proposal, it seems that `reset` and `continues` are arguments to the mode rather than modifiers.

Is it possible to combine *initial* with *continue* or *reset*? I mean something link this `initial ModeA(reset)`. If this is not possible, then I would argue we shouldn't treat *initial* separately.

I really like @oowekyala's suggestion as it is flexible and easily extendable. Also, attributes can be chained if that is needed (`@initial @reset ModeA`)

**Reply by a-sr** [reply]
No, *initial* cannot be combined with *continue*/*reset*. The first one is a modifier of the mode while the others are modifiers of the transition (reaction effect).
```
initial mode A {
  reaction() -> continue(B) {==}
}
mode B {}
```

**Comment by lhstrh** [comment]
Generally, I always like to be able to read programming constructs as if they were prose. In that light, I would say it is more natural to write `reset(modeX)` and `continue(ModeY)` (i.e., "reset to ModeX" and "continue to ModeY"). I think that with this minor syntax adjustment, this would mesh really well with the current reaction interfaces. I also think that this is much more readable (and prominent) than annotations.

When it comes to the labeling of initial states, I think that either maintaining the `initial` keyword or an `@initial` annotation could work well. If we end up using an annotation instead, for the sake of consistency, I think we should also remove the `federated` keyword and turn it into a `@federated` annotation. I personally think these constructs deserve to be front and center though, and we shouldn't be shy about reserving keywords for them.

**Comment by edwardalee** [comment]
I agree with @lhstrh about not being shy about reserving keywords. Right now, our user community is small, so the pain of introducing new keywords won't be big (mostly we just have to modify a few tests).  I'm hoping that situation won't last very long, though, so if we are going to reserve keywords, it's much better to do it sooner rather than later.

**Comment by a-sr** [comment]
I really like the proposal of @lhstrh with the transition type first, also in regard of the interleaved syntax. I adjusted my PR accordingly.

Regarding the annotation syntax, I am a bit reluctant. Personally, I often perceive annotations as of lesser importance (e.g. `@Override`). If we turn these modifiers into annotations, I agree with @lhstrh that it should be done consistently with the entire language, hence, also for *federated* and *main*. However, I am also convinced that for a coordination language like LF we can reserve these keywords for structurally and semantically important constructs. That is why I kept *initial* as a keyword for now. *continue* and *reset*, did not really reached this threshold from my perspective. They seem more common (I can easily imagine scenarios where I would like to have ports with these names in my models). At the same time, I expect 99% of modal models not using either of the modifier because most transitions will use the syntax without keyword, which resets by default. Hence, since the implementation for keyword-less transition modifiers is already there, I think we can keep it like this.

**Comment by cmnrd** [comment]
I initially thought that `initial` is also used in the reaction syntax. Since it actually is a marker on a `mode` much like `main` and `federated` are markers on `reactor`, I would also vote for adding this keyword.

I also think that @lhstrh's suggestion would be a good solution for marking reset and continue transitions.
